### ISSUE TITLE
Zelda - Link Between World Long running issue of Link not running straight

### ISSUE THREAD

**Issue by Br00ty** [citra-emu/citra#5781]
This issue (as far as I know) dates back a few years, but there seems to be a definite issue with (The Legend of Zelda - A Link Between Worlds) you not being able to properly go up/down/left/right in game. You always stray to the left in game instead of staying centered, which makes a lot of the platforming a pain. I almost wonder if there is a way to add an option to be able to change the center of your joystick on Citra, or if its just possible to completely fix that issue. Everything I read on it/watch on it, it's been there since early Citra, and any thoughts on resolution would be greatly appreciated! I will link some references to this issue that I speak of. But I know everyone I have talked to that plays the game experiences the same thing. 

https://youtu.be/DqWqes6DoA4
^^I uploaded the evidence to Youtube as the file was to big.^^

**Comment by RinMaru** [comment]
looks like stick drift

**Comment by Br00ty** [comment]
> looks like stick drift

It looks like it, but it is not unfortunately. It does it on my buddies citra build (we all have the Nightly 1706 build now since they just updated) and quite a few youtube videos. I've also tested it with my other controllers, all cause Link to curve to the left slightly. It's only in ALBW too, all other games on Citra work fine, all go straight. It's so strange. Setting  your circle pad with the "set analog stick" alleviates it a bit since you can adjust your direction, but if you hold the stick up straight on a 8 sided joystick, in reality Link is still curving left.

**Comment by fpi1337** [comment]
Sounds very weird. I also played ALBW (like 2 or 3 qeeks ago) and never had that issue you are reporting. Maybe there's something wrong with your input config?

**Comment by poudink** [comment]
Does it happen if you map the circle pad to a d-pad or keyboard?

**Comment by MiYakuGaming** [comment]
I played the Japanese version of Link Between Worlds and I had the same problem, but on an older build.

**Comment by RokkumanX** [comment]
I can confirm this behavior on the latest Citra Canary release I tested this on (1976) with the European version of the game.

I can't really tell if I noticed this before when I originally played through the game a couple of years ago but now this is all over the place because of this issue, you know what they say...once you see it.

Controllers used:

XBOX 360 Controller
Wii U Pro Controller
Nintendo Switch Pro Controller
Sony PlayStation2 DualShock
Super Famicom.......because why the hell not

Windows 10 1909 Enterprise x64
i7 8700K @stock
NVIDIA 1070
64GB RAM

**Comment by Peduls** [comment]
I have this issue as well on Nightly and Canary. It also affects Tri Force Heroes.

**Comment by Peduls** [comment]
> 
> 
> Does it happen if you map the circle pad to a d-pad or keyboard?

I can confirm it on keyboard. Link drifts to the left when moving towards the top of the screen and to the right when going down.

**Comment by Peduls** [comment]
I also just noticed it happens when moving left and right as well. This was captured using Citra's default controller config for keyboard.

https://user-images.githubusercontent.com/34750422/126091269-54833fce-794b-4470-9f43-efa141470d74.mp4

**Comment by Peduls** [comment]
and here's the same issue in Tri Force Heroes.

https://user-images.githubusercontent.com/34750422/126241701-99802eb4-e53b-423f-aa9e-01a20867a84a.mp4

**Comment by Br00ty** [comment]
Its not the case for real hardware tho. Does not happen on 3DS. this is emulator specific for sure

**Comment by fpi1337** [comment]
@KevinCathcart Thank you for this information! I think that will help the devs a lot. And yeah, it makes sense what you've said.

@Br00ty But he just said he could reproduce it on real hardware. Maybe that issue is not that obvious on real hardware because of the circle pad and the described dead zone it has. That goes with your workaround, setting the analog stick instead of manually assigning directions.

**Comment by KevinCathcart** [comment]
Yeah on PC many gamepads have both horizontal and vertical deadzones, in addition to the center deadzone. Unlike the center deadzone, these deadzones deactivate if you move out of them, and only reactivate when you return to center. This while making it easier to hit the four cardinal directions perfectly, but also allows you to rotate the stick without having the jerky response that they would have if the deadzones were always active. The actual implementation of this can be arbitrarilly complex, but the idea still stands. 

The player movement is off by perhaps 3 to 6 degrees which is often inside of those deadzones, making it hard to subconsciously counteract this this as you must overshoot to deactivate the deadzone and then return to the correct angle. On the original 3ds it just feels like you must have pushed the pad a bit to the side and automatically subconsciously adjust.

I've still no idea what causes this. My new best guess is that they might have limited links movement to a finite number of directions, perhaps 64 or 128, and they wrote code to try to map the pad's position to the nearest such direction, and managed to have an off-by-one error in that code. I don't actually feel like trying to test if this theory is correct.

Whatever the cause, on real hardware it is very hard to tell this is happening unless you specifically look for it, so this error could easily have slipped though quality control testing. And honestly it is pretty harmless on the physical console, but obnoxious on an emulator.

**Comment by itsotter** [comment]
This issue drives me crazy. Following for updates and hoping for an eventual hack to fix it.
